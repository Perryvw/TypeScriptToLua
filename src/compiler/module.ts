import { SourceNode } from "source-map";
import * as ts from "typescript";
import { escapeString, unescapeLuaString } from "../LuaPrinter";

/**
 * Source code of a single input Lua file.
 * May be constructed from transpiled `.ts` source files, or from real `.lua` files.
 */
export interface Module {
    fileName: string;
    isBuilt: boolean;
    source: SourceNode;
    sourceFiles?: ts.SourceFile[];
}

export type ModuleDependencyResolver = (request: string, position: ts.ReadonlyTextRange) => string | { error: string };

export function buildModule(module: Module, dependencyResolver: ModuleDependencyResolver) {
    if (module.isBuilt) return;
    module.isBuilt = true;
    replaceResolveRequests(module.source, dependencyResolver);
}

function replaceResolveRequests(rootNode: SourceNode, dependencyResolver: ModuleDependencyResolver) {
    let currentPosition = 0;

    function replaceInString(source: string) {
        const matches = source.matchAll(/__TS__Resolve\((".*?")\)/g);
        for (const match of [...matches].reverse()) {
            const request = unescapeLuaString(match[1]);
            const pos = currentPosition + match.index!;
            const end = pos + match[0].length;
            const result = dependencyResolver(request, { pos, end });
            const replacement =
                typeof result === "string"
                    ? escapeString(result)
                    : `--[[ ${request} ]] error(${escapeString(result.error)})`;

            source = source.slice(0, match.index) + replacement + source.slice(end);
        }

        return source;
    }

    function walkSourceNode(node: SourceNode, parent: SourceNode) {
        for (const child of node.children as Array<SourceNode | string>) {
            if (typeof child === "object") {
                walkSourceNode(child, node);
            } else {
                if (child.includes("__TS__Resolve")) {
                    parent.children = [replaceInString(parent.toString()) as any];
                }

                currentPosition += child.length;
            }
        }
    }

    walkSourceNode(rootNode, rootNode);
}
