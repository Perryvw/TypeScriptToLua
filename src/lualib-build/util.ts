import * as tstl from "..";

export function isExportTableDeclaration(node: tstl.Node): node is tstl.VariableDeclarationStatement & { left: [] } {
    return tstl.isVariableDeclarationStatement(node) && isExportTable(node.left[0]);
}

export function isExportTable(node: tstl.Node): node is tstl.Identifier {
    return tstl.isIdentifier(node) && node.text === "____exports";
}

export type ExportTableIndex = tstl.TableIndexExpression & { index: tstl.StringLiteral };
export function isExportTableIndex(node: tstl.Node): node is ExportTableIndex {
    return tstl.isTableIndexExpression(node) && isExportTable(node.table) && tstl.isStringLiteral(node.index);
}

export function isExportAlias(
    node: tstl.Node,
): node is tstl.VariableDeclarationStatement & { right: [ExportTableIndex] } {
    return tstl.isVariableDeclarationStatement(node) && node.right !== undefined && isExportTableIndex(node.right[0]);
}

export type ExportAssignment = tstl.AssignmentStatement & { left: [ExportTableIndex] };
export function isExportAssignment(node: tstl.Node): node is ExportAssignment {
    return tstl.isAssignmentStatement(node) && isExportTableIndex(node.left[0]);
}

export function isRequire(node: tstl.Node) {
    return (
        tstl.isVariableDeclarationStatement(node) &&
        node.right &&
        tstl.isCallExpression(node.right[0]) &&
        tstl.isIdentifier(node.right[0].expression) &&
        node.right[0].expression.text === "require"
    );
}

export function isImport(node: tstl.Node, importNames: Set<string>) {
    return tstl.isVariableDeclarationStatement(node) && importNames.has(node.left[0].text);
}

export function isExportsReturn(node: tstl.Node) {
    return (
        tstl.isReturnStatement(node) &&
        tstl.isIdentifier(node.expressions[0]) &&
        node.expressions[0].text === "____exports"
    );
}
