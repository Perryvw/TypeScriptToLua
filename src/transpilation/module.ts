import { SourceNode } from "source-map";
import * as ts from "typescript";
import { escapeString, unescapeLuaString } from "../LuaPrinter";

export interface Module {
    request: string;
    isBuilt: boolean;
    source: SourceNode;
    sourceFiles?: ts.SourceFile[];
}

export type ModuleDependencyResolver = (request: string) => string | { error: string };

export function buildModule(module: Module, dependencyResolver: ModuleDependencyResolver) {
    if (module.isBuilt) return;
    module.isBuilt = true;
    replaceResolveRequests(module.source, dependencyResolver);
}

function replaceResolveRequests(rootNode: SourceNode, dependencyResolver: ModuleDependencyResolver) {
    function replaceInString(source: string) {
        return source.replace(/__TS__Resolve\((".*?")\)/g, (_, match) => {
            const request = unescapeLuaString(match);
            const replacement = dependencyResolver(request);
            return typeof replacement === "string"
                ? escapeString(replacement)
                : `--[[ ${request} ]] error(${escapeString(replacement.error)})`;
        });
    }

    function walkSourceNode(node: SourceNode, parent: SourceNode) {
        for (const child of node.children as Array<SourceNode | string>) {
            if (typeof child === "object") {
                walkSourceNode(child, node);
            } else if (child.includes("__TS__Resolve")) {
                parent.children = [replaceInString(parent.toString()) as any];
            }
        }
    }

    walkSourceNode(rootNode, rootNode);
}
