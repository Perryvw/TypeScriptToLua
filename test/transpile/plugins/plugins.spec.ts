import * as path from "path";
import * as util from "../../util";

test("printer", () => {
    util.testModule``
        .setOptions({ luaPlugins: [{ name: path.join(__dirname, "printer.ts") }] })
        .tap(builder => expect(builder.getMainLuaCodeChunk()).toMatch("-- Custom printer plugin:"));
});

test("printer in bundle", () => {
    const { transpiledFiles } = util.testBundle`
        import "./otherfile";

        const foo = "foo";
    `
        .addExtraFile(
            "otherfile.ts",
            `
                const bar = "bar";
            `
        )
        .setOptions({ luaPlugins: [{ name: path.join(__dirname, "printer.ts") }] })
        .expectToHaveNoDiagnostics()
        .getLuaResult();

    expect(transpiledFiles).toHaveLength(1);
    const lua = transpiledFiles[0].lua!;

    expect(lua).toContain("-- Custom printer plugin: main.lua");
    expect(lua).toContain("-- Custom printer plugin: otherfile.lua");
});

test("visitor", () => {
    util.testFunction`
        return false;
    `
        .setOptions({ luaPlugins: [{ name: path.join(__dirname, "visitor.ts") }] })
        .expectToEqual(true);
});

test("visitor using super", () => {
    util.testFunction`
        return "foo";
    `
        .setOptions({ luaPlugins: [{ name: path.join(__dirname, "visitor-super.ts") }] })
        .expectToEqual("bar");
});

test("passing arguments", () => {
    util.testFunction`
        return {};
    `
        .setOptions({ luaPlugins: [{ name: path.join(__dirname, "arguments.ts"), option: true }] })
        .expectToEqual({ name: path.join(__dirname, "arguments.ts"), option: true });
});

test("statement comments", () => {
    util.testFunction`
        let foo = "bar";
        foo = "baz";
        while (true) {}
    `
        .setOptions({ luaPlugins: [{ name: path.join(__dirname, "add-comments.ts") }] })
        .expectLuaToMatchSnapshot();
});

// https://github.com/TypeScriptToLua/TypeScriptToLua/issues/1013
test.each(["namespace", "module"])("%s with TS transformer plugin", moduleOrNamespace => {
    util.testModule`
        import { ns } from "module";
        export const result = ns.returnsBool();
    `
        .addExtraFile(
            "module.ts",
            `
            export ${moduleOrNamespace} ns {
                export function returnsBool() {
                    return false;
                }
            }
        `
        )
        .setOptions({ plugins: [{ transform: path.join(__dirname, "transformer-plugin.ts") }] })
        .expectNoExecutionError();
});
