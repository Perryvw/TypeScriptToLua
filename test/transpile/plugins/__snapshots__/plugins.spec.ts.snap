// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`statement comments 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    --This comment on variable declaration was added by a plugin!
    --- This one too!
    local foo = \\"bar\\"
    -- This trailing comment was also added by a plugin!
    --- Example luadoc comment
    ---@param paramName ParamClass
    foo = \\"baz\\"
    --[[ This plugin can also (kinda) create multiline comments.
    -- Line 2
    --]]
end
return ____exports"
`;
